### PR TITLE
[main] Add Nuget vulnerabilities output

### PR DESCRIPTION
**Pull request by YuliiaKovalova**
Add Nuget vulnerabilities output on install/update/update check
https://github.com/dotnet/templating/issues/5863

### CODE REVIEW

File: src/Cli/Microsoft.TemplateEngine.Cli/TemplatePackageDisplayer.cs [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+using Microsoft.TemplateEngine.Cli.TabularOutput;
+using Microsoft.TemplateEngine.Edge;
+using Microsoft.TemplateEngine.Edge.Settings;
+
+namespace Microsoft.TemplateEngine.Cli
+{
+    /// <summary>
+    /// The class is responsible for displaying template package manipulation results to console.
+    /// </summary>
+    internal class TemplatePackageDisplayer
```
Comment (vlada-shubina) [inline src/Cli/Microsoft.TemplateEngine.Cli/TemplatePackageDisplayer.cs:21]:
```suggestion
    internal class TemplatePackageDisplay
```

We already have similar class naming

File: src/Cli/Microsoft.TemplateEngine.Cli/TemplatePackageDisplayer.cs [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+    {
+        private readonly IDictionary<int, string> _vulnerabilitySeverityToColorMap = new Dictionary<int, string>
```
Comment (vlada-shubina) [inline src/Cli/Microsoft.TemplateEngine.Cli/TemplatePackageDisplayer.cs:23]:
```suggestion
        private readonly IReadOnlyDictionary<int, string> _vulnerabilitySeverityToColorMap = new Dictionary<int, string>
```

File: src/Cli/Microsoft.TemplateEngine.Cli/TemplatePackageDisplayer.cs [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+        {
+            { 0, LocalizableStrings.TemplatePackageCoordinator_VulnerabilitySeverity_Low.Cyan() },
+            { 1, LocalizableStrings.TemplatePackageCoordinator_VulnerabilitySeverity_Moderate.Blue() },
+            { 2, LocalizableStrings.TemplatePackageCoordinator_VulnerabilitySeverity_High.Yellow() },
+            { 3, LocalizableStrings.TemplatePackageCoordinator_VulnerabilitySeverity_Critical.Red() },
+        };
+
+        private readonly bool _xterm = Environment.GetEnvironmentVariable("TERM")?.Contains("xterm") ?? false;
```
Comment (vlada-shubina) [inline src/Cli/Microsoft.TemplateEngine.Cli/TemplatePackageDisplayer.cs:31]: can this be moved to `AnsiConsole`?

File: src/Cli/Microsoft.TemplateEngine.Cli/TemplatePackageDisplayer.cs [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+                        break;
+                    case InstallerErrorCode.InvalidPackage:
+                        _reporterError.WriteLine(
+                              string.Format(
+                                  LocalizableStrings.TemplatePackageCoordinator_lnstall_Error_InvalidPackage,
+                                  packageToInstall).Bold().Red());
+                        break;
+
+                    case InstallerErrorCode.VulnerablePackage:
+                        {
```
Comment (vlada-shubina) [inline src/Cli/Microsoft.TemplateEngine.Cli/TemplatePackageDisplayer.cs:178]: you can do something like
var vulnerabilities = (result as InstallResult)?.Vulnerabilities ?? (result as UpdateResult)?.Vulnerabilities and remove duplicate code.

**Review by vlada-shubina — CHANGES_REQUESTED** [review]
looks good in general, couple of things: 
- examples should be changed to match other examples
- consider handling links in `AnsiConsole`
- some wording suggestions

File: src/Cli/Microsoft.TemplateEngine.Cli/TemplatePackageDisplayer.cs (hunk 4 above)
Comment (YuliiaKovalova) [reply src/Cli/Microsoft.TemplateEngine.Cli/TemplatePackageDisplayer.cs:178]: depending on the command install/update we print different messages.
https://github.com/dotnet/sdk/pull/32002/files#:~:text=1mFailed%20to%20install%20testPackage%2C%20due%20to%20detected%20vulnerabilities

https://github.com/dotnet/sdk/pull/32002/files#:~:text=1mThe%20package%20testPackage%20wasn%27t

so I don't see the way to remove these pieces of code

File: src/Cli/Microsoft.TemplateEngine.Cli/Commands/create/TemplateCommand.cs [hunk 5]
```diff
@@ -151,6 +151,7 @@ internal async Task<NewCommandStatus> InvokeAsync(ParseResult parseResult, Cance
             TemplateInvoker invoker = new TemplateInvoker(_environmentSettings, () => Console.ReadLine() ?? string.Empty);
             TemplatePackageCoordinator packageCoordinator = new TemplatePackageCoordinator(_environmentSettings, _templatePackageManager);
             TemplateConstraintManager constraintManager = new TemplateConstraintManager(_environmentSettings);
+            TemplatePackageDisplay templatePackageDisplayer = new TemplatePackageDisplay(Reporter.Output, Reporter.Error);
```
Comment (vlada-shubina) [inline src/Cli/Microsoft.TemplateEngine.Cli/Commands/create/TemplateCommand.cs:154]:
```suggestion
            TemplatePackageDisplay templatePackageDisplay = new TemplatePackageDisplay(Reporter.Output, Reporter.Error);
```

File: src/Cli/Microsoft.TemplateEngine.Cli/TemplatePackageCoordinator.cs [hunk 6]
```diff
@@ -28,15 +25,15 @@ internal class TemplatePackageCoordinator
         private readonly TemplatePackageManager _templatePackageManager;
         private readonly TemplateConstraintManager _constraintsManager;
         private readonly HostSpecificDataLoader _hostSpecificDataLoader;
+        private readonly TemplatePackageDisplay _templatePackageDisplayer;
```
Comment (vlada-shubina) [inline src/Cli/Microsoft.TemplateEngine.Cli/TemplatePackageCoordinator.cs:28]:
```suggestion
        private readonly TemplatePackageDisplay _templatePackageDisplay;
```

File: src/Cli/Microsoft.TemplateEngine.Cli/TemplatePackageDisplay.cs [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+        private void DisplayVulnerabilityInfo(IReadOnlyList<VulnerabilityInfo> vulnerabilities, IReporter reporter)
+        {
+            reporter.Write(Environment.NewLine);
+            foreach (VulnerabilityInfo entry in vulnerabilities)
+            {
+                reporter.WriteLine($"{string.Empty.PadLeft(4)}{_vulnerabilitySeverityToColorMap[entry.Severity].Bold()}:");
+                reporter.WriteLine(
+                    string.Join(
+                        Environment.NewLine,
+                        entry.AdvisoryUris.Select(advisory => $"{string.Empty.PadLeft(8)}{(AnsiConsole.XtermEnabled ? advisory.Url() : advisory)}")));
```
Comment (vlada-shubina) [inline src/Cli/Microsoft.TemplateEngine.Cli/TemplatePackageDisplay.cs:382]: this should be preferably handled in AnsiConsole, base on XtermEnabled value

File: src/Cli/Microsoft.TemplateEngine.Cli/TemplatePackageDisplay.cs [hunk 8]
```diff
@@ -0,0 +1,420 @@
...
+                    break;
+                case InstallerErrorCode.UnsupportedRequest:
+                    _reporterError.WriteLine(
+                        string.Format(
+                            LocalizableStrings.TemplatePackageCoordinator_Update_Error_PackageNotSupported,
+                            result.TemplatePackage.DisplayName).Bold().Red());
+                    break;
+                case InstallerErrorCode.VulnerablePackage:
+                    {
+                        if (result.Vulnerabilities.Any())
```
Comment (vlada-shubina) [inline src/Cli/Microsoft.TemplateEngine.Cli/TemplatePackageDisplay.cs:352]: if the code is `VulnerablePackage` we should print error message

File: src/Cli/Microsoft.TemplateEngine.Cli/TemplatePackageDisplay.cs [hunk 9]
```diff
@@ -0,0 +1,420 @@
...
+                            {
+                                _reporterError.WriteLine(string.Format(
+                                    LocalizableStrings.TemplatePackageCoordinator_Update_Error_VulnerablePackage,
+                                    packageToInstall).Bold().Red());
+                                DisplayVulnerabilityInfo(updateRequest.Vulnerabilities, _reporterError);
+                                _reporterError.WriteLine(string.Format(
+                                    LocalizableStrings.TemplatePackageCoordinator_Update_Error_VulnerablePackageTip,
+                                    packageToInstall,
+                                    SharedOptions.ForceOption.Aliases.First()).Bold());
+                            }
```
Comment (vlada-shubina) [inline src/Cli/Microsoft.TemplateEngine.Cli/TemplatePackageDisplay.cs:198]: `throw new InvalidOperationException("Unexpected result: typeof(result).")`

File: src/Cli/Microsoft.TemplateEngine.Cli/TemplatePackageDisplay.cs [hunk 10]
```diff
@@ -0,0 +1,420 @@
...
+                    case InstallerErrorCode.InvalidPackage:
+                        _reporterError.WriteLine(
+                              string.Format(
+                                  LocalizableStrings.TemplatePackageCoordinator_lnstall_Error_InvalidPackage,
+                                  packageToInstall).Bold().Red());
+                        break;
+
+                    case InstallerErrorCode.VulnerablePackage:
+                        {
+                            if (result is InstallResult installResult && installResult.Vulnerabilities.Any())
```
Comment (vlada-shubina) [inline src/Cli/Microsoft.TemplateEngine.Cli/TemplatePackageDisplay.cs:178]: it's better to use `switch` for this then.

File: src/Cli/Microsoft.TemplateEngine.Cli/LocalizableStrings.resx [hunk 11]
```diff
@@ -270,6 +270,40 @@
   <data name="TemplatePackageCoordinator_Uninstall_Error_GenericError" xml:space="preserve">
     <value>Failed to uninstall {0}, reason: {1}.</value>
   </data>
+  <data name="TemplatePackageCoordinator_Update_Error_VulnerablePackage" xml:space="preserve">
+    <value>The package {0} was not updated due to detected vulnerabilities:</value>
+    <comment>{0} - template package name.</comment>
+  </data>
+  <data name="TemplatePackageCoordinator_Update_Error_VulnerablePackageTip" xml:space="preserve">
+    <value>In order to use this package run:</value>
```
Comment (vlada-shubina) [inline src/Cli/Microsoft.TemplateEngine.Cli/LocalizableStrings.resx:278]:
```suggestion
    <value>In order to update this package, run:</value>
```

**Review by vlada-shubina — APPROVED** [review]

**Pull request merged**